### PR TITLE
feat: Added 'slugify' sprig function

### LINKED ISSUE #10622: Fix: Sticking to DNS label standard in branch_slug (Gitlab PR Generators)
# Summary

When using the "Build" job from Gitlab auto-devops, the image name includes the slugified version of the branch ( using this variable `CI_COMMIT_REF_SLUG`) which is limited to 63 characters. So when trying to use the `branch_slug` key to building the path of the image, Argo won't find it because in my old PR https://github.com/argoproj/argo-cd/pull/9462 I hardcoded the slug max length to 50.

# Motivation

Same motivation as https://github.com/argoproj/argo-cd/pull/9462
In order to follow the DNS label standard as defined in RFC 1123, manifests' name metadata should not exceed 63 and must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character.

Docs: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names

# Proposal

This time after falling into the limit trap, I suggest sticking to DN label standard and keep the limit as 63 characters.

### PR DESCRIPTION
**Pull request by Aym3nTN**
Closes https://github.com/argoproj/argo-cd/issues/10622

### CODE REVIEW

File: applicationset/utils/utils.go [hunk 1]
```diff
@@ -448,6 +450,54 @@ func SanitizeName(name string) string {
...
+		case len(args) - 1:
+			name = arg.(string)
+		case 0:
+			maxSize = arg.(int)
+		case 1:
+			EnableSmartTruncate = arg.(bool)
+		default:
+			log.Errorf("Bad 'slugify' arguments.")
+		}
+	}
```
Comment (crenshaw-dev) [inline applicationset/utils/utils.go:487]: Hm. I'm out of my element with go templating, but why the dynamic parameters?

File: applicationset/utils/utils_test.go [hunk 2]
```diff
@@ -1185,6 +1185,43 @@ func TestNormalizeBitbucketBasePath(t *testing.T) {
 	}
 }
 
+func TestSlugify(t *testing.T) {
```
Comment (crenshaw-dev) [inline applicationset/utils/utils_test.go:1188]: Would you mind also adding a test that demonstrates this function works in an AppSet template?

File: applicationset/utils/utils.go [hunk 3]
```diff
@@ -448,6 +450,54 @@ func SanitizeName(name string) string {
 	return strings.Trim(name, "-.")
 }
 
+// SlugifyName generates a URL-friendly slug from the provided name and additional options.
```
Comment (crenshaw-dev) [inline applicationset/utils/utils.go:453]: Could you update the AppSet docs to describe this function and give some example usage?

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: applicationset/utils/utils.go (hunk 1 above)
Comment (Aym3nTN) [reply applicationset/utils/utils.go:487]: I wanted to give the end user the possibility to omit some arguments and give the chance to add more in the future for backward compatibility; with dynamic params all of the following is correct
- {{ cat branch | slugify }}
- {{ cat branch | slugify 50 }}
- {{ cat branch | slugify 50 false }}

File: applicationset/utils/utils.go (hunk 3 above)
Comment (Aym3nTN) [reply applicationset/utils/utils.go:453]: What about somewhere [here](https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/GoTemplate/#introduction:~:text=An%20additional%20normalize%20function%20makes%20any%20string%20parameter%20usable%20as%20a%20valid%20DNS%20name%20by%20replacing%20invalid%20characters%20with%20hyphens%20and%20truncating%20at%20253%20characters.%20This%20is%20useful%20when%20making%20parameters%20safe%20for%20things%20like%20Application%20names.) ?

File: applicationset/utils/utils.go (hunk 1 above)
Comment (crenshaw-dev) [reply applicationset/utils/utils.go:487]: That's fancy as heck. Okay, sounds good to me! I'm not sure how error handling typically works in go template functions, but should we capture type check failures etc. and report those up as an error? Or is panicking typical for these functions?

Comment (Aym3nTN) [reply applicationset/utils/utils.go:487]: Here is how it throws the error, it's somehow clear that the parameters are wrong

**Comment by Aym3nTN** [comment]
Just documented the function and wrote the tests, please have a look 👀

**Comment by crenshaw-dev** [comment]
Looks like the unit test failure is Real

**Comment by Aym3nTN** [comment]
> Looks like the unit test failure is Real

A bad test case from some another MR was added to `TestGenerateAppsUsingPullRequestGenerator` by mistake, I have removed it.

**Comment by Aym3nTN** [comment]
@crenshaw-dev All checks have passed 🎉

**Comment by Aym3nTN** [comment]
@crenshaw-dev I've re-resolved the conflicts, would you please merge it this time?

**Comment by Aym3nTN** [comment]
@crenshaw-dev in case you've missed my previous comment

> @crenshaw-dev I've re-resolved the conflicts, would you please merge it this time?

**Review by crenshaw-dev — APPROVED** [review]

**Pull request merged**